### ISSUE TITLE
Unsafety checks for "unconst" operations do not trigger in unstable functions

### ISSUE THREAD

**Issue by ecstatic-morse** [rust-lang/rust#72394]
Some operations, such as raw pointer comparisons, require `unsafe` when inside a const context. To my knowledge, none of these operations have been const-stabilized. Accordingly, the [following example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=374960212a8444136ca2de7440bd4a76) fails to compile due to a missing unsafe block.

```rust
#![feature(const_compare_raw_pointers)]
#![feature(const_fn)]

const fn unstable(a: *const i32, b: *const i32) -> bool {
    a == b
}
```

However, the check does not run for `const_unstable` functions. The following [compiles successfully](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ef079890f9f2fddddd6b2c6578be7c9e):

```rust
#![stable(feature = "foo", since = "1.33.0")]
#![feature(staged_api)]
#![feature(const_compare_raw_pointers)]
#![feature(const_fn)]

#[stable(feature = "foo", since = "1.33.0")]
#[rustc_const_unstable(feature = "const_foo", issue = "none")]
const fn unstable(a: *const i32, b: *const i32) -> bool {
    a == b
}
```

This is an easy problem to fix, but the root cause here is poor naming. Specifically, `fn_queries::is_const_fn` should be renamed `is_callable_as_const_fn` and `is_const_fn_raw` should be `is_declared_const_fn`. We have had other issues due to this in the past. cc @rust-lang/wg-const-eval to weigh in on the renaming.

**Comment by RalfJung** [comment]
Yeah I recall being confused by those names in the past, too. Definite :+1: for changing them.

As for concrete proposals, I don't actually remember what these queries are doing, but your suggestions sound reasonable. :)

**Comment by RalfJung** [comment]
Oh, the names are different than in the email I got because you edited.^^

So does `is_callable_as_const_fn` take the context (feature flags etc) into account? The name sounds like it would. What about `is_const_callable_fn`, to make it a bit shorter?

**Comment by ecstatic-morse** [comment]
> Oh, the names are different than in the email I got because you edited.

Damn. I was hoping no one would catch me :smile:. 

`is_const_callable_fn` would indeed take feature flags into account. My original name was `is_stably_const_fn`, but that's not really true, since it would also return `true` if the function was unstable but the correct feature flags were set.
